samples/apps/forum/src/controllers/poll.ts

```diff
@@ -0,0 +1,305 @@
+    @Response<ValidateErrorResponse>(ValidateErrorStatus, "Schema validation error")
+    @Put('all')
+    public submitOpinions(
+        @Body() body: SubmitOpinionsRequest,
+        @Header() authorization: string,
+    ): void {
+        const user = parseAuthToken(authorization)
+
+        for (const [topic, opinion] of Object.entries(body.opinions)) {
+            try {
+                var poll = this.kvPolls.get(topic)
+            } catch (e) {
+                this.setStatus(400)
+                return { message: `Poll with topic '${topic}' does not exist` } as any
+            }
+            if (typeof opinion.opinion !== poll.type) {
+                this.setStatus(400)
+                return { message: `Poll with topic '${topic}' has a different opinion type` } as any
+            }      
+            poll.opinions[user] = opinion.opinion
+            this.kvPolls.set(topic, poll)
+        }
+
+        this.setStatus(204)
+    }
+
+    @SuccessResponse(200, "Poll data")
+    @Response<ErrorResponse>(404, "Poll data could not be returned because no poll with the given topic exists")
+    @Response<ValidateErrorResponse>(ValidateErrorStatus, "Schema validation error")
+    @Get()
+    public getPoll(
+        @Query() topic: string,
+        @Header() authorization: string,
+    ): GetPollResponse {
+        const user = parseAuthToken(authorization)
+
+        if (!this.kvPolls.has(topic)){
+            this.setStatus(404)
+            return { message: "Poll does not exist" } as any
+        }
+
+        this.setStatus(200)
+        return this._getPoll(user, topic)
+    }
+    
+    @SuccessResponse(200, "Poll data")
+    @Response<ValidateErrorResponse>(ValidateErrorStatus, "Schema validation error")
+    @Get('all')
+    public getPolls(
+        @Header() authorization: string,
+    ): GetPollsResponse {
+        const user = parseAuthToken(authorization)
+
+        let response: GetPollsResponse = { polls: {} }
+
+        for (const topic of this._getTopics()) {
+            response.polls[topic] = this._getPoll(user, topic)
+        }
+
+        this.setStatus(200)
+        return response
+    }
+
+    _getTopics(): string[] {
+        try {
+            return this.kvTopics.get(this.kvTopicsKey)
+        } catch (e) {
+            return []
+        }
+    }
+
+    _getPoll(user: string, topic: string): GetPollResponse {
+        try {
+            var poll = this.kvPolls.get(topic)
+        } catch (e) {
+            throw new Error(`Poll with topic '${topic}' does not exist`)
+        }
+
+        const opinionCountAboveThreshold = Object.keys(poll.opinions).length >= MINIMUM_OPINION_THRESHOLD
+
+        let response: GetPollResponse
+        // TODO can repetition be avoided while maintaining type checking?
+        if (poll.type == "string") {
+            response = {
```

Comment:
Seems like at least response = { type: poll.type } should hoist out without trouble?